modules/dists/BlockDist.chpl

```diff
@@ -1538,9 +1538,62 @@ where this.sparseLayoutType == unmanaged DefaultDist &&
   return true;
 }
 
+proc BlockArr.canDoOptimizedSwap(other) {
+  var domsMatch = true;
+  var domShapesMatch = true;
+
+  if this.dom != other.dom { // no need to check if this is true
+    if domsMatch {
+      for param i in 0..this.dom.rank-1 {
+        if this.dom.whole.dim(i) != other.dom.whole.dim(i) {
+          if this.dom.whole.dim(i).size != other.dom.whole.dim(i).size {
+            domShapesMatch = false;
+          }
+          domsMatch = false;
+        }
+      }
+    }
+  }
+
+  if domsMatch {
+    // distributions must be equal, too
+    return this.dom.dist.dsiEqualDMaps(other.dom.dist);
+  }
+  // we can handle different domains if:
+  // their shapes match, and
+  // they also match their respective distributions' boundingBoxes
+  else if domShapesMatch {
+    // TODO: this rule can be relaxed a bit.  ie. arrays on the following
+    // domains have matching shaped and can be swapped with a pointer swap, but
+    // that's not what happens because of this line is only checking for a very
+    // specific case.
+    //
+    // var d1 = {0..9 by 2 align 0} dmapped Block({1..10});
+    // var d2 = {0..9 by 2 align 1} dmapped Block({1..10});
+    return this.dom.dist.boundingBox == this.dom.whole &&
+           other.dom.dist.boundingBox == other.dom.whole;
+  }
+  return false;
+}
+
+// <=> implementation uses Reflection to check whether this resolves, so
+// it is OK for this to not resolve without debugOptimizedSwap.
+proc BlockArr.doiOptimizedSwap(other) where debugOptimizedSwap {
```

Comment:
I must be misunderstanding something:  This looks as though if `debugOptimizedSwap` is true, we will never do an optimized swap (because this routine returns `false`).

Reply:
If `debugOptimizedSwap` is true, this overload will be called for cases where we are swapping between different types. See that the other main overload of this function has `other: this.type` as an argument.

The only reason for this overload is to provide some debugging output without actually making the main overload a generic one, thinking that fully generic functions are additional load on resolution. But that may be a premature optimization and can cause more maintenance overhead than its worth. What do you think?

Reply:
I see now, thanks for the explanation.  It's a little thing, but if we keep this, I'd move it after the "actual" doiOptimizedSwap() so that someone searching for the routine (from the top of the file) or reading the code from top to bottom will come across it first rather than this overload.  I'd probably also update the comment to more proactively state why it's here, e.g., "The purpose of this overload is to provide debugging output in the event that `debugOptimizedSwap` is on and the main routine doesn't resolve (e.g., due to a type, stridability, or rank mismatch in the `other` argument).  When `debugOptimizedSwap` is off, this overload will be ignored due to its `where` clause."

My (potentially incorrect) sense is that `where` clauses also add resolution load (moreso than formal argument checks I think Michael has said), but I couldn't say how a generic function whose body returns false compares to an overload with a false where clause.  In any case, I don't have an objection to the routine or its where clause if the comment had made its role a bit clearer (the current comment focuses more on mechanism and doesn't relate it to the "real" routine at all).

Reply:
I'll keep the function in with the adjustments you suggest.

My thinking was that if we have a fully generic version of the function that returns false for some types, that'd mean we instantiate that function for many different type combinations only to return false and/or give a debug message. With this kind of implementation we create a doiOptimizedSwap only if there is a type match and what we need to do is some runtime checks. (Unless you explicitly set `debugOptimizedSwap`)

Do any of these cause noticeable impact for swap? probably not given it is not an integral part of the interface :)